rsync_to_rdisc.py

```diff
@@ -69,124 +64,156 @@ def send_email(sender, receivers, subject, text, attachment=None):
+    transferred_dic = {}
+
+    if os.path.isfile(str(wkdir) + "transferred_runs.txt"):
+        with open("{}/transferred_runs.txt".format(wkdir), 'r') as runs:
+            for line in runs:
+                transferred_dic[line.rstrip()] = ""
+    else:
+        new_file = open(str(wkdir) + "transferred_runs.txt", "w")
+        new_file.close()
+
+    """ Get folders to be transfer from HPC """
+    client = SSHClient()
+    client.load_host_keys(settings.host_keys)
+    client.load_system_host_keys()
+    try:
+        client.connect(settings.server[0], username=settings.user)
+        hpc_server = settings.server[0]
+    except OSError:
+        try:
+            client.connect(settings.server[1], username=settings.user)
+            hpc_server = settings.server[1]
+        except OSError:
+            make_mail(" and ".join(settings.server), "lost_hpc", run_file=run_file)
+            sys.exit("connection to HPC transfernodes are lost")
+
+    to_be_transferred = {}
+    folder_dic = settings.folder_dic
+    for folder in folder_dic:
+        stdin, stdout, stderr = client.exec_command("ls {}".format(folder_dic[folder]["input"]))
+        folders = stdout.read().decode("utf8")
+        for item in folders.split():
+            combined = "{0}_{1}".format(item.split()[-1], folder)
+            if combined not in transferred_dic:
+                to_be_transferred[item.split()[-1]] = folder
+
+    """Rsync folders from HPC to bgarray"""
+    log = settings.log
+    errorlog = settings.errorlog
+    temperror = settings.temperror
+    date = str(datetime.datetime.now()).split(".")[0]
+    bgarray_log_file = "{bgarray}/{log}".format(bgarray=settings.bgarray, log=log)
+    remove_run_file = True
+
+    for run in to_be_transferred:
+        continue_rsync = True
+        folder = settings.folder_dic[to_be_transferred[run]]
+        action = ("rsync -rahuL --stats {user}@{server}:{path}{run} {output}/ \
+            1>> {bgarray}/{log} 2>> {bgarray}/{errorlog} 2> {temperror}".format(
+                user=settings.user,
+                server=hpc_server,
+                path=folder["input"],
+                run=run,
+                output=folder["output"],
+                bgarray=settings.bgarray,
+                log=log,
+                errorlog=errorlog,
+                temperror=temperror
+            )
+        )
+
+        missing = []
+        for check_file in folder["files_required"]:
+            if check_file:
+                stdin, stdout, stderr = client.exec_command((
+                    "[[ -f {0}{1}/{2} ]] && echo \"Present\" "
+                    "|| echo \"Absent\""
+                ).format(folder["input"], run, check_file))
+                status = stdout.read().decode("utf8").rstrip()
+                if status == "Absent":
+                    missing.append(check_file)
+
+        if missing:
+            if 'continue_without_email' in folder and folder["continue_without_email"]:
```

Comment:
What is the benefit of having both checks?
Would `if folder["continue_without_email"]:` be sufficient?

Reply:
The idea was to have this parameter optional in de dict. Therefor you need to first check whether the parameter is in dict before you can check its value.